Async transfers

sdijoseph opened the pull request:
Draft PR for adding async transfer feature mentioned in issue #144 

This involved some pretty major changes to UsbContext and UsbDevice (See commit a34825a5aa4e1b891d5ef5496e10143fac6ba3b8). I was attempting to follow the suggestion mentioned in the [libusb API documentation](https://libusb.sourceforge.io/api-1.0/group__libusb__asyncio.html#details) (Under "Using an event handling thread"). The change in this commit should also prevent users from using UsbDevices whose context has been disposed (either explicitly or by GC).

Known issues:
- ~~On endpoint writes where an interface must be claimed, unless I put an `await Task.Delay(1)` before the interface is released, libusb_close hangs if the device is the last open one on the UsbContext.~~ Fixed now.
- Just generally need a lot more testing. I'm working on getting one of the [suggested libusb test devices](https://github.com/libusb/libusb/wiki/FAQ#user-content-Do_you_have_a_list_of_known_good_devices_to_test_libusb) to do so.

Features not yet included:
- Should probably allow retries on transfer submission, since this is mentioned as one of the perks of the async model in the libusb documenation.
- ~~I'm thinking it might be nice to have a static collection track all created UsbContexts, and dispose of all of them in a static ShutDown method, allowing users to just call that on application exit. It's a lot more important to dispose UsbContexts now, because if a device is left open then the event handling thead will hang and leave the process open.~~ No longer the case, as the event handling thread is a background thread.

mcuee commented:
@qmfrederik 

This is one major missing feature for the main branch. Please review this PR when you have some time. Thanks.

mcuee commented:
@sdijoseph 

One of the giithub action build (libusbdotnet for Windows) failed. Please take a look. Thanks.

sdijoseph commented:
> @sdijoseph
> 
> One of the giithub action build (libusbdotnet for Windows) failed. Please take a look. Thanks.

Latest commit should fix it, I'm developing on Linux so I forgot ValueTuple needs to be included as a package for .NET 4.5

mcuee commented:
> Latest commit should fix it, I'm developing on Linux so I forgot ValueTuple needs to be included as a package for .NET 4.5

Yes, the issue is fixed. Just wondering if you can fix BenchmarkCon to use the new async transfer so that I can carry out testing using it. If the test passes, then it will give people some good confidence to this PR. Thanks.

mcuee commented:
@sdijoseph 
The updated example seems to work fine, with the Benchmark FW (0x04d8:0xfa2e). 
But there is a minor issue that I have to use CTRL-C to exit the application. Strange. I am using WinUSB driver here.

```
PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> .\src\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH

Done! (Note: need to use CTRL-C to exit)

```

There is no such issue for libusbdotnet v2 branch.
```
PS C:\work\libusb\libusbdotnet\libusbdotnet_v2> .\stage\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
Read  :8 ErrorCode:Success
Write :8 ErrorCode:Success
Data  :ABCDEFGH
Read  :8 ErrorCode:Success
Write :8 ErrorCode:Success
Data  :ABCDEFGH
Read  :8 ErrorCode:Success
Write :8 ErrorCode:Success
Data  :ABCDEFGH
Read  :8 ErrorCode:Success
Write :8 ErrorCode:Success
Data  :ABCDEFGH
Read  :8 ErrorCode:Success
Write :8 ErrorCode:Success
Data  :ABCDEFGH

Done!(I can use any key to exit, including RETURN).

```

sdijoseph commented:
> @sdijoseph The updated example seems to work fine, with the Benchmark FW (0x04d8:0xfa2e). But there is a minor issue that I have to use CTRL-C to exit the application. Strange. I am using WinUSB driver here.
> 
> ```
> PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> .\src\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
> Read  :8 Error:Success
> Write :8 Error:Success
> Data  :ABCDEFGH
> Read  :8 Error:Success
> Write :8 Error:Success
> Data  :ABCDEFGH
> Read  :8 Error:Success
> Write :8 Error:Success
> Data  :ABCDEFGH
> Read  :8 Error:Success
> Write :8 Error:Success
> Data  :ABCDEFGH
> Read  :8 Error:Success
> Write :8 Error:Success
> Data  :ABCDEFGH
> 
> Done! (Note: need to use CTRL-C to exit)
> ```
> 
> There is no such issue for libusbdotnet v2 branch.
> 
> ```
> PS C:\work\libusb\libusbdotnet\libusbdotnet_v2> .\stage\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
> Read  :8 ErrorCode:Success
> Write :8 ErrorCode:Success
> Data  :ABCDEFGH
> Read  :8 ErrorCode:Success
> Write :8 ErrorCode:Success
> Data  :ABCDEFGH
> Read  :8 ErrorCode:Success
> Write :8 ErrorCode:Success
> Data  :ABCDEFGH
> Read  :8 ErrorCode:Success
> Write :8 ErrorCode:Success
> Data  :ABCDEFGH
> Read  :8 ErrorCode:Success
> Write :8 ErrorCode:Success
> Data  :ABCDEFGH
> 
> Done!(I can use any key to exit, including RETURN).
> ```

No, this is expected. One of the known issues with this PR is that unless I put an `await Task.Delay(1)` after the final transfer but before closing the device, libusb_close hangs and prevents the process from terminating (interestingly, `Thread.Sleep(1)` does not prevent the hang). 

I need to investigate it some more, I'm suspecting multi-threading shenanigans.

sdijoseph commented:
So the libusb_close hanging issue was because the transfer callback was run on the event handling thread, but then the task completion was also continued on the event handling thread. This eventually caused Close to be called on the event handling thread as well, which is why it hung until the thread was yielded with Task.Delay. 

The solution was to add TaskCreationOptions.RunContinuationsAsynchronously to the task completion source of the transfer. Unfortunately, this option wasn't added until .NET 4.6, so I had to bump the framework version for now. Please let me know if anyone knows how to get the same behavior in 4.5.

I also made the event handling thread a background thread so it doesn't block the process from exiting if it isn't joined properly.

> Yes, the issue is fixed. Just wondering if you can fix BenchmarkCon to use the new async transfer so that I can carry out testing using it. If the test passes, then it will give people some good confidence to this PR. Thanks.

I should be getting my test board soon, then I can start working on this.

mcuee commented:
Yes the latest git is good, tested under Windows 11.

```
PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> dotnet test .\src\LibUsbDotNet.Tests\LibUsbDotNet.Tests.csproj
  Determining projects to restore...
  All projects are up-to-date for restore.
  LibUsbDotNet -> C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet\bin\Debug\netcoreapp3.1\LibUsbDotNet.
  dll
  LibUsbDotNet.Tests -> C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Tests\bin\Debug\net6.0\LibUsbDo
  tNet.Tests.dll
Test run for C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Tests\bin\Debug\net6.0\LibUsbDotNet.Tests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.5.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     4, Skipped:     0, Total:     4, Duration: 5 ms - LibUsbDotNet.Tests.dll (net6.0)
PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> .\src\examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH

Done! (Hit return and it exits normally)
```

mcuee commented:
https://learn.microsoft.com/en-us/dotnet/framework/migration-guide/versions-and-dependencies

@sdijoseph 
.NET Framework 4.6 is included in VS 2015 and works under Windows Vista/7/8/8.1/10/11. 
.NET Framework 4.5 is included in VS 2012 and works under Windows Vista/7/8/8.1/10/11.

So I do not see it as a real issue to bump the required version from 4.5 to 4.6.

sdijoseph commented:
> @sdijoseph The updated example seems to work fine, with the Benchmark FW (0x04d8:0xfa2e). But there is a minor issue that I have to use CTRL-C to exit the application. Strange. I am using WinUSB driver here.

@mcuee I got the FX3 board. Can you link me the FW you are using (or the name if it is one included in the fx3 sdk)? I tried using [this one](https://github.com/mcuee/libusbk/tree/master/BmFW/CYPRESS_FX3/BMFW_BulkSourceSink) but it appears to be the wrong one, since it forces me to set the alt interface to 1 and also spits out weird data:
```
Read  :1024 Error:Success
Write :8 Error:Success
Data  : 


 !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~��������������������������������������������������������������������������������������������������������������������������������


 !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~��������������������������������������������������������������������������������������������������������������������������������


 !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~��������������������������������������������������������������������������������������������������������������������������������


 !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~��������������������������������������������������������������������������������������������������������������������������������
```



> https://learn.microsoft.com/en-us/dotnet/framework/migration-guide/versions-and-dependencies
> 
> @sdijoseph .NET Framework 4.6 is included in VS 2015 and works under Windows Vista/7/8/8.1/10/11. .NET Framework 4.5 is included in VS 2012 and works under Windows Vista/7/8/8.1/10/11.
> 
> So I do not see it as a real issue to bump the required version from 4.5 to 4.6.

Ok, it's fine with me if it's fine with you.

mcuee commented:
> @mcuee I got the FX3 board. Can you link me the FW you are using (or the name if it is one included in the fx3 sdk)? I tried using [this one](https://github.com/mcuee/libusbk/tree/master/BmFW/CYPRESS_FX3/BMFW_BulkSourceSink) but it appears to be the wrong one, since it forces me to set the alt interface to 1 and also spits out weird data:

The FW examples here work with libusbK kBench program and they are supposed to work with libusbdotnet benchmarkcon application.
https://github.com/mcuee/libusbk/tree/master/BmFW/CYPRESS_FX3

The default FX3 SDK examples like BulkLoop, BulkSrcSink, IsoLoop and IsoSrcSink are working with libusbK kBench application (with `notestselect` option). They are supposed to work with libusbdotnet benchmarkcon application as well, with `notestselect` option.

But I believe the isochronous example may not work too well with libusbdotnet benchmarkcon yet.

mcuee commented:
Example run for the Cypress FX3 SDK BulkLoop example with libusbdotnet benchmarkcon example (from V2 branch).
```
PS C:\work\libusb\libusbdotnet\libusbdotnet_v2> .\stage\BenchmarkCon\bin\Release\net6.0\BenchmarkCon.exe notestselect list
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 1C7A:0575 Egis Technology Inc. - EgisTec Touch Fingerprint Sensor
2. 04B4:00F0 Cypress Semiconductor Corp. - FX3

Select device (1-2) :2

Benchmark device 04B4:00F0 opened..
Loop Test Information
        Vid / Pid       : 04B4h / 00F0h
        Interface #     : 00h
        Alt Interface # : 00h
        Priority        : Normal
        Buffer Size     : 4096
        Buffer Count    : 1
        Display Refresh : 1000 (ms)
        Transfer Timeout: 5000 (ms)
        Retry Count     : 0
        Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
Bulk Write (Ep01h) max packet size: 1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
Avg. Bytes/s: 52049768.60 Transfers: 12635 Bytes/s: 52049768.60
Avg. Bytes/s: 51942415.21 Transfers: 25387 Bytes/s: 51836482.66
Avg. Bytes/s: 52405051.67 Transfers: 38534 Bytes/s: 53322137.52
Avg. Bytes/s: 52227435.84 Transfers: 51284 Bytes/s: 51697875.89
Avg. Bytes/s: 52432236.23 Transfers: 64306 Bytes/s: 53254656.63
Avg. Bytes/s: 52137899.63 Transfers: 76699 Bytes/s: 50662176.82
Avg. Bytes/s: 51873909.53 Transfers: 88996 Bytes/s: 50285837.06
Avg. Bytes/s: 51650687.04 Transfers: 101228 Bytes/s: 50082674.65
Avg. Bytes/s: 51604127.38 Transfers: 113749 Bytes/s: 51230768.74
Avg. Bytes/s: 51408546.55 Transfers: 125990 Bytes/s: 49659603.01
Avg. Bytes/s: 51254625.22 Transfers: 138150 Bytes/s: 49712458.92
qstopped Ep81h thread.
stopped Ep01h thread.
Loop Test Information
        Vid / Pid       : 04B4h / 00F0h
        Interface #     : 00h
        Alt Interface # : 00h
        Priority        : Normal
        Buffer Size     : 4096
        Buffer Count    : 1
        Display Refresh : 1000 (ms)
        Transfer Timeout: 5000 (ms)
        Retry Count     : 0
        Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
        Total Bytes     : 616419328
        Total Transfers : 150493
        Avg. Bytes/sec  : 51145973.49
        Elapsed Time    : 12.05 seconds

Bulk Write (Ep01h) max packet size: 1024
        Total Bytes     : 616419328
        Total Transfers : 150493
        Avg. Bytes/sec  : 51145804.59
        Elapsed Time    : 12.05 seconds

Press any key to exit..

```

mcuee commented:
@sdijoseph 

I just tried the example from your PR and it works fine with the FX3 SDK BulkLoop example (`VID:PID=0x04b4:00f0`).

```
PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> dotnet build .\src\Examples\Read.Write.Async\Read.Write.Async.csproj -c Release
MSBuild version 17.4.1+2eac9157a for .NET
  Determining projects to restore...
C:\Program Files\dotnet\sdk\7.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(28,5): war
ning NETSDK1138: The target framework 'net5.0' is out of support and will not receive security updates in the future. P
lease refer to https://aka.ms/dotnet-core-support for more information about the support policy. [C:\work\libusb\libusb
dotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\Read.Write.Async.csproj::TargetFramework=net5.0]
  Restored C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\Read.Write.Async.csproj (in 120
   ms).
  Restored C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet\LibUsbDotNet.csproj (in 180 ms).
C:\Program Files\dotnet\sdk\7.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(28,5): war
ning NETSDK1138: The target framework 'net5.0' is out of support and will not receive security updates in the future. P
lease refer to https://aka.ms/dotnet-core-support for more information about the support policy. [C:\work\libusb\libusb
dotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\Read.Write.Async.csproj::TargetFramework=net5.0]
  LibUsbDotNet -> C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet\bin\Release\netcoreapp3.1\LibUsbDotNe
  t.dll
  Read.Write.Async -> C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\bin\Release\net6.0\R
  ead.Write.Async.dll
  Read.Write.Async -> C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\bin\Release\net5.0\R
  ead.Write.Async.dll

Build succeeded.

C:\Program Files\dotnet\sdk\7.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(28,5): war
ning NETSDK1138: The target framework 'net5.0' is out of support and will not receive security updates in the future. P
lease refer to https://aka.ms/dotnet-core-support for more information about the support policy. [C:\work\libusb\libusb
dotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\Read.Write.Async.csproj::TargetFramework=net5.0]
C:\Program Files\dotnet\sdk\7.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(28,5): war
ning NETSDK1138: The target framework 'net5.0' is out of support and will not receive security updates in the future. P
lease refer to https://aka.ms/dotnet-core-support for more information about the support policy. [C:\work\libusb\libusb
dotnet\libusbdotnet_pr193\src\Examples\Read.Write.Async\Read.Write.Async.csproj::TargetFramework=net5.0]
    2 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.63

PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> .\src\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH

Done!

```

mcuee commented:
BTW, libusbK FX3 examples do not seem to work with libusbdotnet benchmarkcon as of now. They do work with libusbK kBench application.

```
PS C:\work\libusb\libusbdotnet\libusbdotnet_v2> .\stage\BenchmarkCon\bin\Release\net6.0\BenchmarkCon.exe list
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 1C7A:0575 Egis Technology Inc. - EgisTec Touch Fingerprint Sensor
2. 04B4:00FE Cypress Semiconductor Corp. - FX3 BULKLOOP

Select device (1-2) :2

Benchmark device 04B4:00FE opened..
err:failed locating EP80h!
err:failed creating transfer param!
Press any key to exit..

PS C:\work\libusb\libusbdotnet\libusbdotnet_v2> cd C:\libusbK-dev-kit_release\
PS C:\libusbK-dev-kit_release> .\kBench.exe list
device-count=1
1. FX3 BULKLOOP (USB\VID_04B4&PID_00FE\5&586B51A&0&13) [WinUSB]
Select device (1-1) :1

opened FX3 BULKLOOP (USB\VID_04B4&PID_00FE\5&586B51A&0&13)..
Loop Test Information
        Driver          : WinUSB
        Vid / Pid       : 04B4h / 00FEh
        DevicePath      : \\?\usb#vid_04b4&pid_00fe#5&586b51a&0&13#{290b097a-88d0-8c45-af12-74df5776749c}
        Device Speed    : High
        Interface #     : 00h
        Alt Interface # : 01h
        Num Endpoints   : 2
        Priority        : 0
        Read Size       : 4096
        Write Size      : 4096
        Buffer Count    : 1
        Display Refresh : 1000 (ms)
        Transfer Timeout: 5000 (ms)
        Retry Count     : 0
        Verify Data     : Off

Bulk Read (Ep81h) Maximum Packet Size:1024
Bulk Write (Ep01h) Maximum Packet Size:1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
Avg. Bytes/s: 107241472.00 Transfers: 26182 Bytes/s: 107241472.00
Avg. Bytes/s: 107458560.00 Transfers: 52470 Bytes/s: 107675648.00
Avg. Bytes/s: 107834026.67 Transfers: 78980 Bytes/s: 108584960.00
Avg. Bytes/s: 107941888.00 Transfers: 105412 Bytes/s: 108265472.00
Avg. Bytes/s: 107760844.80 Transfers: 131544 Bytes/s: 107036672.00
Avg. Bytes/s: 107293957.45 Transfers: 157588 Bytes/s: 104996283.46
Avg. Bytes/s: 107009906.78 Transfers: 183492 Bytes/s: 105318712.75
Avg. Bytes/s: 107042745.36 Transfers: 209878 Bytes/s: 107278396.53
Avg. Bytes/s: 106688624.67 Transfers: 234840 Bytes/s: 103801372.59
Avg. Bytes/s: 106069699.13 Transfers: 259762 Bytes/s: 100571933.00
Avg. Bytes/s: 105785298.88 Transfers: 284892 Bytes/s: 102932480.00
stopped Ep81h thread.   ExitCode=0
stopped Ep01h thread.   ExitCode=0
Loop Test Information
        Driver          : WinUSB
        Vid / Pid       : 04B4h / 00FEh
        DevicePath      : \\?\usb#vid_04b4&pid_00fe#5&586b51a&0&13#{290b097a-88d0-8c45-af12-74df5776749c}
        Device Speed    : High
        Interface #     : 00h
        Alt Interface # : 01h
        Num Endpoints   : 2
        Priority        : 0
        Read Size       : 4096
        Write Size      : 4096
        Buffer Count    : 1
        Display Refresh : 1000 (ms)
        Transfer Timeout: 5000 (ms)
        Retry Count     : 0
        Verify Data     : Off

Bulk Read (Ep81h) Maximum Packet Size:1024
        Total Bytes     : 636764160
        Total Transfers : 155460
        Avg. Bytes/sec  : 52926952.04
        Elapsed Time    : 12.03 seconds

Bulk Write (Ep01h) Maximum Packet Size:1024
        Total Bytes     : 636764160
        Total Transfers : 155460
        Avg. Bytes/sec  : 52926952.04
        Elapsed Time    : 12.03 seconds

Press any key to exit..

```

mcuee commented:
I also tested the Benchmark FW on Microchip PIC18F87J90 PIM (0x04d8:0xFA2E, VID/PID licensed from Microchip).
FW: https://github.com/mcuee/libusbk/tree/master/BmFW/MCP

```
mcuee@UbuntuSwift3:~/build/libusbdotnet_pr193$ ./src/Examples/Read.Write.Async/bin/Release/net6.0/Read.Write.Async 
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH
Read  :8 Error:Success
Write :8 Error:Success
Data  :ABCDEFGH

Done!

```

mcuee commented:
@MatCDev and @GreenShadeZhang

Just wondering if you have some time to review this PR as well. Thanks. With this PR, one of the major missing features of the master branch (v3) branch will be addressed. v3 branch is still the way to go for libusbdotnet project.

mcuee commented:
@sdijoseph 

BTW, currently LibUsbDotNet.Generator is broken. This is not caused by the PR as it is already broken in the master branch. Just wondering if you can fix it as well. Thanks.
* https://github.com/LibUsbDotNet/LibUsbDotNet/issues/201
```
C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Generator\FunctionVisitor.cs(195,42): error CS0246: The
 type or namespace name 'Comment' could not be found (are you missing a using directive or an assembly reference?) [C:\
work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Generator\LibUsbDotNet.Generator.csproj]
C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Generator\StructVisitor.cs(275,49): error CS0246: The t
ype or namespace name 'Comment' could not be found (are you missing a using directive or an assembly reference?) [C:\wo
rk\libusb\libusbdotnet\libusbdotnet_pr193\src\LibUsbDotNet.Generator\LibUsbDotNet.Generator.csproj]
```

mcuee commented:
The other two issues related to this PR are related to the `Read.Isochronous` and `BenchmarkCon` which I believe will be fixed later.
```
C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\Examples\Read.Isochronous\ReadIsochronous.cs(184,42): error CS0246:
The type or namespace name 'UsbTransferQueue' could not be found (are you missing a using directive or an assembly refe
rence?) [C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\Examples\Read.Isochronous\Read.Isochronous.csproj]
C:\work\libusb\libusbdotnet\libusbdotnet_pr193\src\BenchmarkCon\BenchmarkConsole.cs(126,20): error CS0246: The type or
namespace name 'UsbTransfer' could not be found (are you missing a using directive or an assembly reference?) [C:\work\
libusb\libusbdotnet\libusbdotnet_pr193\src\BenchmarkCon\BenchmarkCon.csproj]
```

sdijoseph commented:
I improved the Read.WriteAsync example to better demonstrate the transfers are being sent asynchronously.

Run with FX3 0x04b4 0x00f0 BulkLoop FW:

```
Write transfer #7 completed @ 18.2954 ms with Error-Success Length-7 Data-13.8698
Read transfer #2 completed @ 18.2982 ms with Error-Success Length-7 Data-16.1269
Write transfer #3 completed @ 18.3019 ms with Error-Success Length-7 Data-13.8211
Read transfer #5 completed @ 18.3041 ms with Error-Success Length-7 Data-16.1269
Write transfer #2 completed @ 18.3071 ms with Error-Success Length-7 Data-13.8096
Read transfer #0 completed @ 18.3141 ms with Error-Success Length-7 Data-16.1269
Write transfer #4 completed @ 18.3156 ms with Error-Success Length-7 Data-13.8311
Read transfer #3 completed @ 18.3181 ms with Error-Success Length-7 Data-16.1269
Read transfer #1 completed @ 18.3212 ms with Error-Success Length-7 Data-16.1269
Write transfer #6 completed @ 18.3307 ms with Error-Success Length-7 Data-13.8603
Write transfer #1 completed @ 18.333 ms with Error-Success Length-7 Data-13.7839
Read transfer #6 completed @ 18.3418 ms with Error-Success Length-7 Data-16.1269
Write transfer #0 completed @ 18.3453 ms with Error-Success Length-7 Data-11.7474
Write transfer #5 completed @ 18.3486 ms with Error-Success Length-7 Data-13.8501
Read transfer #9 completed @ 18.6946 ms with Error-Success Length-6 Data-16.126
Write transfer #8 completed @ 18.6948 ms with Error-Success Length-7 Data-16.1079
Write transfer #9 completed @ 18.6949 ms with Error-Success Length-6 Data-16.126
Read transfer #7 completed @ 18.6968 ms with Error-Success Length-7 Data-16.1269
Read transfer #8 completed @ 18.7 ms with Error-Success Length-7 Data-16.1269
Read transfer #4 completed @ 22.7009 ms with Error-Success Length-7 Data-16.1269

Done!
```

mcuee commented:
Test with PIC18F87J50 running the Microchip PIC Benchmark FW (0x04d8:0xfa2e)
https://github.com/mcuee/libusbk/tree/master/BmFW/MCP

The results look good but I need to hit CTRL-C to exit

```
PS C:\work\libusb\libusbdotnet\libusbdotnet_pr193> .\src\Examples\Read.Write.Async\bin\Release\net6.0\Read.Write.Async.exe
Read transfer #1 completed @ 21.8049 ms with Error-Success Length-7 Data-17.2557
Read transfer #2 completed @ 21.8049 ms with Error-Success Length-7 Data-17.2557
Read transfer #0 completed @ 21.8111 ms with Error-Success Length-6 Data-17.255
Read transfer #3 completed @ 21.8307 ms with Error-Success Length-7 Data-17.2557
Write transfer #0 completed @ 21.9013 ms with Error-Success Length-6 Data-7.5065
Write transfer #3 completed @ 21.9275 ms with Error-Success Length-7 Data-17.1861
Write transfer #1 completed @ 21.9324 ms with Error-Success Length-7 Data-17.1336
Write transfer #2 completed @ 21.9363 ms with Error-Success Length-7 Data-17.1686
Read transfer #4 completed @ 22.7439 ms with Error-Success Length-7 Data-17.2557
Read transfer #5 completed @ 22.7453 ms with Error-Success Length-7 Data-17.2557
Write transfer #5 completed @ 22.7466 ms with Error-Success Length-7 Data-17.2085
Write transfer #4 completed @ 22.7469 ms with Error-Success Length-7 Data-17.1973
Write transfer #6 completed @ 22.7469 ms with Error-Success Length-7 Data-17.2188
Read transfer #6 completed @ 22.7532 ms with Error-Success Length-7 Data-17.2557
Write transfer #8 completed @ 22.7539 ms with Error-Success Length-7 Data-17.2394
Write transfer #9 completed @ 22.7541 ms with Error-Success Length-7 Data-17.2557
Read transfer #7 completed @ 22.7543 ms with Error-Success Length-7 Data-17.2557
Write transfer #7 completed @ 22.7543 ms with Error-Success Length-7 Data-17.2275
Read transfer #8 completed @ 22.755 ms with Error-Success Length-7 Data-17.2557
Read transfer #9 completed @ 22.7562 ms with Error-Success Length-7 Data-17.2557

Done!
(need to use CTRL-C to exit)
```

sdijoseph commented:
> Test with PIC18F87J50 running the Microchip PIC Benchmark FW (0x04d8:0xfa2e) https://github.com/mcuee/libusbk/tree/master/BmFW/MCP
> 
> The results look good but I need to hit CTRL-C to exit

Ok I fixed it again with the latest commit, adding the condition of UsbContext.IsDisposing was the culprit.

I'm not sure how big a deal it is that threads will be created and destroyed pretty often when only one device is being opened and closed at a time on a given context.

mcuee commented:
@sdijoseph 
Thanks. It seems to me this PR works now pretty well. Just wondering if you can rebase to the master branch. Thanks.

As for the `BenchmarkCon` example, it is nice to have as a good testing tool to test this PR. But it is okay to leave it to the next PR if it is too much work for you. And we can work with normal bulk and interrupt transfer first and worry about Isochronous transfer later. Same for `Read.Isochronous` example which can be dealt later.

sdijoseph commented:
> @sdijoseph Thanks. It seems to me this PR works now pretty well. Just wondering if you can rebase to the master branch. Thanks.
> 
> As for the `BenchmarkCon` example, it is nice to have as a good testing tool to test this PR. But it is okay to leave it to the next PR if it is too much work for you. And we can work with normal bulk and interrupt transfer first and worry about Isochronous transfer later. Same for `Read.Isochronous` example which can be dealt later.

A couple things:

- I deleted the SyncTransfer class just now because it was only called for synchronous UsbEndpointBase.Transfers to control and isosynchronous endpoints, except the call for iso transfers always passed 0 as the iso packet size so that wouldn't have worked. Synchronous control transfers already have a method, so I removed the class.
- I have a transfer cancellation PR in the works, but I would prefer that to be separate, as it should undergo further testing.
- This PR adds async bulk, interrupt, and control transfers. I'll get to iso when I have time, it doesn't seem super complicated looking at the documentation.

For the BenchmarkCon example, it seems like most of the code hasn't been touched in 13 years, and it's a very different style than I'm used to. It would be my preference to add a benchmarking component to the Avalonia application I proposed in #197 . Let me know what you think.

Given all this, are you ready to move forward with merging this? If so, I'll do an interactive rebase to clean up the history and then force push to the branch.

sdijoseph commented:
> I have sent an email to @qmfrederik and asked him to review this PR. He is probably busy with work. Let's give him one more month. If I still do not hear from him after one month, I will merge this PR. Is that okay with you?

Sure, it's all good with me. I did the rebase and marked the PR as ready for review.

mcuee commented:
It is still good, tested under Ubuntu Linux 20.04.

```
mcuee@UbuntuSwift3:~/build/libusb/libusbdotnet_pr193$ dotnet build ./src/Examples/Read.Write.Async/Read.Write.Async.csproj -c Release
MSBuild version 17.5.1+f6fdcf537 for .NET
  Determining projects to restore...
  Restored /home/mcuee/build/libusb/libusbdotnet_pr193/src/Examples/Read.Write.Async/Read.Write.Async.csproj (in 61 ms).
  Restored /home/mcuee/build/libusb/libusbdotnet_pr193/src/LibUsbDotNet/LibUsbDotNet.csproj (in 142 ms).
  LibUsbDotNet -> /home/mcuee/build/libusb/libusbdotnet_pr193/src/LibUsbDotNet/bin/Release/net6.0/LibUsbDotNet.dll
  LibUsbDotNet -> /home/mcuee/build/libusb/libusbdotnet_pr193/src/LibUsbDotNet/bin/Release/net7.0/LibUsbDotNet.dll
  Read.Write.Async -> /home/mcuee/build/libusb/libusbdotnet_pr193/src/Examples/Read.Write.Async/bin/Release/net6.0/Read.Write.Async.dll
  Read.Write.Async -> /home/mcuee/build/libusb/libusbdotnet_pr193/src/Examples/Read.Write.Async/bin/Release/net7.0/Read.Write.Async.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.08
mcuee@UbuntuSwift3:~/build/libusb/libusbdotnet_pr193$ ./src/Examples/Read.Write.Async/bin/Release/net7.0/Read.Write.Async 
Write transfer #3 completed @ 9.5284 ms with Error-Success Length-6 Data-6.3576
Write transfer #0 completed @ 9.531 ms with Error-Success Length-6 Data-4.9689
Read transfer #1 completed @ 9.5335 ms with Error-Success Length-6 Data-6.4433
Read transfer #0 completed @ 9.5363 ms with Error-Success Length-6 Data-6.4433
Write transfer #1 completed @ 9.5391 ms with Error-Success Length-6 Data-6.2817
Read transfer #2 completed @ 9.5421 ms with Error-Success Length-6 Data-6.4433
Write transfer #2 completed @ 9.547 ms with Error-Success Length-6 Data-6.3159
Write transfer #4 completed @ 9.5493 ms with Error-Success Length-6 Data-6.3649
Write transfer #5 completed @ 10.035 ms with Error-Success Length-6 Data-6.3802
Read transfer #3 completed @ 10.0394 ms with Error-Success Length-6 Data-6.4433
Write transfer #7 completed @ 10.0403 ms with Error-Success Length-6 Data-6.4079
Write transfer #8 completed @ 10.043 ms with Error-Success Length-6 Data-6.4311
Read transfer #4 completed @ 10.0489 ms with Error-Success Length-6 Data-6.4433
Write transfer #6 completed @ 10.0497 ms with Error-Success Length-6 Data-6.3932
Read transfer #5 completed @ 10.0508 ms with Error-Success Length-6 Data-6.4433
Read transfer #6 completed @ 10.0751 ms with Error-Success Length-6 Data-6.4433
Read transfer #7 completed @ 10.0772 ms with Error-Success Length-6 Data-6.4433
Write transfer #9 completed @ 10.0789 ms with Error-Success Length-6 Data-6.4433
Read transfer #8 completed @ 10.0802 ms with Error-Success Length-6 Data-6.4433
Read transfer #9 completed @ 10.0814 ms with Error-Success Length-6 Data-6.4433

Done!

```

mcuee commented:
@sdijoseph 

In order to make the testing of PR #216 easier, I have created https://github.com/LibUsbDotNet/LibUsbDotNet/tree/development branch. Please retarget this PR to the development branch and I will merge this PR to the development branch. Thanks.

mcuee commented:
@sdijoseph

On a second thought, this is not really necessary.

qmfrederik commented on src/LibUsbDotNet/LibUsb/AsyncTransfer.cs:
```diff
@@ -19,80 +19,94 @@
+            transferId);
 
-        int transferId = 0;
+        return await transferCompletion.Task.ConfigureAwait(false);
```
nit: you can slightly simplify this code by making the method return `transferCompletion.Task` without awaiting it, e.g.:

```csharp
    public static Task<(Error error, int transferLength)> TransferAsync(
        DeviceHandle device,
        byte endPoint,
        EndpointType endPointType,
        IntPtr buffer,
        int offset,
        int length,
        int timeout,
        int isoPacketSize)
{
    ...

    return transferCompletion.Task;
}
```

qmfrederik commented on src/LibUsbDotNet/LibUsb/UsbEndpointReader.Async.cs:
```diff
@@ -0,0 +1,100 @@
+        /// <returns>
+        /// Tuple of (<see cref="Error"/> error, <see cref="int"/> transferLength). error is <see cref="Error.Success"/> on success.
+        /// </returns>
+        public virtual async Task<(Error error, int transferLength)> ReadAsync(byte[] buffer, int timeout)
```
You could simplify this as:

```csharp
public virtual Task<(Error error, int transferLength)> ReadAsync(byte[] buffer, int timeout) => this.ReadAsync(buffer, 0, buffer.Length, timeout);
```

qmfrederik commented:
@sdijoseph Thank you for this PR!  Apologies for taking so long to review this (and @mcuee for pinging me!).  I've skimmed over the PR and it looks good to me.  The only nit I have is w.r.t. `return await [task]`; you can often just make the method sync and do `return [task]`, it simplifies the code a bit.

sdijoseph commented:
> @sdijoseph Thank you for this PR! Apologies for taking so long to review this (and @mcuee for pinging me!). I've skimmed over the PR and it looks good to me. The only nit I have is w.r.t. `return await [task]`; you can often just make the method sync and do `return [task]`, it simplifies the code a bit.

I actually already did this, along with changing some of the code to use Span across sync apis and Memory across async apis more consistently. Originally I was going to put these changes in a separate PR, but I've been using these changes for a couple of weeks in my projects now, so I think they can go into this PR. The changes are in commit e5484d43ac6dde8bac70649778ce3ddd56c6db4f

mcuee commented:
@sdijoseph 

Thanks for the updates. I have merged this PR.

Please rebase PR #216 and then @qmfrederik will review it.

If you have further improvement to the async feature, you can always raise new PRs. Thanks a lot for your great work!
